checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py

```diff
@@ -137,30 +138,28 @@ def to_fragments(self) -> fragments.Fragments:
     return result
 
 
-def _hashable_slices(slices: Index) -> tuple[HashableSlice, ...]:
-  return tuple([(s.start, s.stop, s.step) for s in slices])
-
-
-def _create_replica_counts_builder(
-    arr: jax.Array,
-) -> Callable[[jax.Shard], int]:
-  """Produces a mapping from addressable shards to global replication count."""
-
-  counts = collections.defaultdict(int)
-  for index in arr.sharding.devices_indices_map(arr.shape).values():
```

Comment:
Note that devices_indices_map is cached so you only do the work once not everytime like your PR description states.

It would be nice to have a more concrete benchmark which proves your hypothesis. This has been used for a long time without any significant problem so I am suspicious of your claim.

Reply:
Fair point re caching `devices_indices_map`. Are you aware of existing code that already causes every process to call `devices_indices_map` for ~every jax.Array?

We've generally been fighting to shave seconds off the startup time for large-scale training jobs, so I'd hate to introduce that dependency on mesh size. Caching aside, I do think the numbers above are realistic in terms of time added upon startup -- e.g. PAXML GPT checkpoints 114 `jax.Array`s.

Reply:
> Are you aware of existing code that already causes every process to call devices_indices_map for ~every jax.Array?
Every jax.Array should be fine but my point is that we should be hitting the cache if the mesh and specs are the same for all those arrays which they usually are.

But I am fine with this change as long as you don't use JAX internal APIs at all. That includes all APIs with `_` in front of them.

Reply:
Afaict the cache is keyed on both the sharding and the global shape, but that should help cut it down a fair bit indeed.

Reply:
Actually, the microbenchmark in the PR description already represents the best case wrt `devices_indices_map`'s cache -- all but the first call are hits. So merely traversing the resulting map and allocating small objects for hashable_slices is what would take hundreds of milliseconds for every checkpoint when running a workload like PAXML GPT (~100 saved arrays) on a mesh of 8k devices.

That being said, I'm now more convinced that the number of `(sharding, global_shape)` pairs seen in practice will be small. So extending the caching to `_sharding_num_replicas(sharding, global_shape)` and implementing it using the public API `sharding.devices_indices_map(global_shape)` is a straightforward solution. Once shardings other than `jax.sharding.NamedSharding` are deprecated we could go back and get rid of `devices_indices_map`.

I updated the PR accordingly -- PTAL @cpgaffney1 !